### PR TITLE
fix: add missing chaintip property

### LINKED ISSUE #3074: Chainstate endpoint has missing `chainTip` property
### Context
Bee 1.6.2, macOS

### Summary
There is a [`/chainstate`](https://docs.ethswarm.org/debug-api/#tag/Status/paths/~1chainstate/get) endpoint that is supposed to report the postage syncing state during postage syncing. If I understand correctly the `chainTip` property should return where the syncing is at and I guess it is the `totalAmount` which it should reach. The problem is that the `chainTip` property is never returned, at least for me it was never returned during postage sycning, neither after it was synced.

## Task
Add the field to the actual API response and return it from the current blockheight of the chain backend that we are using.

### PR DESCRIPTION
**Pull request by swarmHaseeb**
### Checklist

- [ ] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md)
- [ ] My change requires a documentation update and I have done it
- [ ] I have added tests to cover my changes.

### Description
This PR fixes #3074 assigned on July 13th 2022.

#### Motivation and context (Optional)
This is my first commit.

### Screenshots (if appropriate):

---
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/3086)

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -1465,7 +1463,6 @@ golang.org/x/sys v0.0.0-20210603081109-ebe580a85c40/go.mod h1:oPkhp1MJrh7nUepCBc
 golang.org/x/sys v0.0.0-20210615035016-665e8c7367d1/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
 golang.org/x/sys v0.0.0-20210616094352-59db8d763f22/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
 golang.org/x/sys v0.0.0-20210630005230-0f9fa26af87c/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
-golang.org/x/sys v0.0.0-20210816183151-1e6c022a8912/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
```
Comment (acud) [inline go.sum:1468]: please revert the changes to `go.sum` if they are not needed by the changeset

File: pkg/api/postage_test.go [hunk 2]
```diff
@@ -405,6 +410,9 @@ func TestChainState(t *testing.T) {
 		ts, _, _, _ := newTestServer(t, testServerOptions{
```
Comment (acud) [inline pkg/api/postage_test.go:410]: this is a strange test case :) not sure what it is for...

**Review by acud — COMMENTED** [review]
Nice :) congrats on your first PR :)

File: go.sum (hunk 1 above)
Comment (swarmHaseeb) [reply go.sum:1468]: i think we can ignore these changes during merge.

File: pkg/api/postage_test.go (hunk 2 above)
Comment (swarmHaseeb) [reply pkg/api/postage_test.go:410]: should i do anything with it?

**Review by mrekucci — COMMENTED** [review]
> Reviewed 3 of 4 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3086)* status: all files reviewed, 1 unresolved discussion (waiting on @acud)

___
*[`pkg/api/postage.go` line 340 at r2](https://reviewable.io/reviews/ethersphere/bee/3086#-N75RmIU-mWhT23tdKdi:-N75RmIU-mWhT23tdKdj:bsu5thh) ([raw file](https://github.com/ethersphere/bee/blob/dc87cba6315f5c0c04a974905c6553c208c3a1d0/pkg/api/postage.go#L340)):*
> ```Go
> 
> type chainStateResponse struct {
> 	ChainTip     uint64         `json:"chainTip"`     // ChainTip (block height)
> ```

Nitpick: missing period at the end.
___
*[`pkg/api/postage.go` line 372 at r2](https://reviewable.io/reviews/ethersphere/bee/3086#-N75SHI5Cy10EG9--6qg:-N75SHI5Cy10EG9--6qh:bo4xlx7) ([raw file](https://github.com/ethersphere/bee/blob/dc87cba6315f5c0c04a974905c6553c208c3a1d0/pkg/api/postage.go#L372)):*
> ```Go
> 	logger := tracing.NewLoggerWithTraceID(r.Context(), s.logger)
> 	state := s.batchStore.GetChainState()
> 	chainTip, err := s.chainBackend.BlockNumber(context.Background())
> ```

The `t.Context()` should also be passed here. I'd recommend creating a new variable `ctx := r.Context()` before the `logger` and pass it where context is required.
___
*[`pkg/api/postage.go` line 374 at r2](https://reviewable.io/reviews/ethersphere/bee/3086#-N75SuBG3qI2raYEthI6:-N75SuBG3qI2raYEthI7:b-l98613) ([raw file](https://github.com/ethersphere/bee/blob/dc87cba6315f5c0c04a974905c6553c208c3a1d0/pkg/api/postage.go#L374)):*
> ```Go
> 	chainTip, err := s.chainBackend.BlockNumber(context.Background())
> 	if err != nil {
> 		logger.Debugf("chainstate: block number: %v", err)
> ```

The debug message should, in most cases, copy the error message (plus the error after the semicolon).

**Comment by swarmHaseeb** [comment]
> 

Hi, umm can you please explain the last suggestion again?

File: pkg/api/postage.go [hunk 3]
```diff
@@ -363,10 +366,18 @@ func (s *Service) reserveStateHandler(w http.ResponseWriter, _ *http.Request) {
 }
 
 // chainStateHandler returns the current chain state.
-func (s *Service) chainStateHandler(w http.ResponseWriter, _ *http.Request) {
+func (s *Service) chainStateHandler(w http.ResponseWriter, r *http.Request) {
+	logger := tracing.NewLoggerWithTraceID(r.Context(), s.logger)
 	state := s.batchStore.GetChainState()
-
+	chainTip, err := s.chainBackend.BlockNumber(context.Background())
```
Comment (istae) [inline pkg/api/postage.go:372]: better to use `r.Context()` here instead of a new background

**Comment by aloknerurkar** [comment]
> > 
> 
> Hi, umm can you please explain the last suggestion again?
> 
> Ref:
> 
> ```
> 	if err != nil {
> 		logger.Debugf("chainstate: block number: %v", err)
> ```
> 
> Also, for the snippet below, I can just pass the context without creating additional variable if that's ok.
> 
> ```
> 	state := s.batchStore.GetChainState()
> 	chainTip, err := s.chainBackend.BlockNumber(r.Context())
> ```

I think what @mrekucci means is that there are multiple context arguments. So lets have 1 variable which is initialized using the `r.Context()`. And then use this variable everywhere. Because this was not done, we ended up in a situation where we use 2 different contexts.

**Review by istae — CHANGES_REQUESTED** [review]

**Review by swarmHaseeb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3086)* status: 1 of 201 files reviewed, 2 unresolved discussions (waiting on @acud, @istae, and @mrekucci)

___
*[`pkg/api/postage.go` line 372 at r2](https://reviewable.io/reviews/ethersphere/bee/3086#-N6wAsDBkGVPyaEAyD0i-r2-372:-N7GeogKBBYppabpZQaH:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/dc87cba6315f5c0c04a974905c6553c208c3a1d0/pkg/api/postage.go#L372)):*
<details><summary><i>Previously, istae (istae) wrote…</i></summary><blockquote>

better to use `r.Context()` here instead of a new background
</blockquote></details>

Already Done.

File: pkg/api/postage_test.go (hunk 2 above)
Comment (acud) [reply pkg/api/postage_test.go:410]: please remove it

**Review by swarmHaseeb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3086)* status: 1 of 201 files reviewed, 2 unresolved discussions (waiting on @acud, @istae, and @mrekucci)

___
*[`pkg/api/postage_test.go` line 410 at r2](https://reviewable.io/reviews/ethersphere/bee/3086#-N6wAsDBkGVPyaEAyD0j-r2-410:-N7Ka4q30l2GQsWRUFJj:bikz5b6) ([raw file](https://github.com/ethersphere/bee/blob/dc87cba6315f5c0c04a974905c6553c208c3a1d0/pkg/api/postage_test.go#L410)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

please remove it
</blockquote></details>

Done, empty test case removed.

File: go.sum (hunk 1 above)
Comment (istae) [reply go.sum:1468]: yeah, please revert

File: pkg/api/postage.go [hunk 4]
```diff
@@ -363,10 +365,18 @@ func (s *Service) reserveStateHandler(w http.ResponseWriter, _ *http.Request) {
...
-
+	chainTip, err := s.chainBackend.BlockNumber(r.Context())
+	if err != nil {
+		logger.Debugf("chainstate: block number: %v", err)
+		logger.Error("chainstate: block number unavailable")
+		jsonhttp.InternalServerError(w, nil)
+		return
+	}
 	jsonhttp.OK(w, chainStateResponse{
+		ChainTip:     chainTip,
```
Comment (istae) [inline pkg/api/postage.go:379]: because this is an API change, the docs have to be updated, along with the openapi specs.

**Review by istae — CHANGES_REQUESTED** [review]

File: pkg/api/postage.go (hunk 4 above)
Comment (acud) [reply pkg/api/postage.go:379]: @istae the property was in the docs but missing from the endpoint :shrug:

**Review by swarmHaseeb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3086)* status: 0 of 201 files reviewed, 3 unresolved discussions (waiting on @acud, @istae, and @mrekucci)

___
*[`pkg/api/postage.go` line 379 at r5](https://reviewable.io/reviews/ethersphere/bee/3086#-N6wAsDBkGVPyaEAyD0i-r5-379:-N7LsPH4B3skT4SaWWcH:bl63o2z) ([raw file](https://github.com/ethersphere/bee/blob/5bdaa2bd82634be0b27db738b811d08aa05bbb84/pkg/api/postage.go#L379)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

@istae the property was in the docs but missing from the endpoint :shrug: 
</blockquote></details>

property already present in docs.

File: pkg/api/postage.go (hunk 4 above)
Comment (acud) [reply pkg/api/postage.go:379]: https://docs.ethswarm.org/debug-api/#tag/Status/paths/~1chainstate/get

File: go.sum (hunk 1 above)
Comment (swarmHaseeb) [reply go.sum]: done

File: pkg/api/postage.go [hunk 5]
```diff
@@ -9,6 +9,7 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"github.com/ethersphere/bee/pkg/tracing"
```
Comment (aloknerurkar) [inline pkg/api/postage.go:12]: I commented on the other PR as well, you should be using gofmt. Please follow the coding style guide.

**Review by swarmHaseeb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/3086)* status: 0 of 201 files reviewed, 4 unresolved discussions (waiting on @acud, @aloknerurkar, @istae, and @mrekucci)

___
*[`pkg/api/postage.go` line 379 at r5](https://reviewable.io/reviews/ethersphere/bee/3086#-N6wAsDBkGVPyaEAyD0i-r5-379:-N7UEkZ83Ue4O6jPr8o9:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/5bdaa2bd82634be0b27db738b811d08aa05bbb84/pkg/api/postage.go#L379)):*
<details><summary><i>Previously, acud (acud) wrote…</i></summary><blockquote>

https://docs.ethswarm.org/debug-api/#tag/Status/paths/~1chainstate/get
</blockquote></details>

Done.
___
*[`pkg/api/postage.go` line 12 at r6](https://reviewable.io/reviews/ethersphere/bee/3086#-N6wAsDBkGVPyaEAyD0i-r6-12:-N7UEiqrDGr5-cvLMYjb:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/1679261a109e686d4a5c6a4e25ce6cb753f705cb/pkg/api/postage.go#L12)):*
<details><summary><i>Previously, aloknerurkar wrote…</i></summary><blockquote>

I commented on the other PR as well, you should be using gofmt. Please follow the coding style guide.
</blockquote></details>

Done.

**Review by acud — APPROVED** [review]

**Review by aloknerurkar — APPROVED** [review]

**Review by istae — APPROVED** [review]
